File: android/app/src/main/java/com/pixel/reimburse/SMSReceiver.java
package com.pixel.reimburse;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.telephony.SmsMessage;
import android.util.Log;

public class SMSReceiver extends BroadcastReceiver {
    private static final String TAG = "SMSMonitor";
    private static final java.util.regex.Pattern TRANSACTION_PATTERN = java.util.regex.Pattern.compile(
        "rs\\.|inr|debited|spent|transaction|vpa|paid to|credited|upi|merchant|purchased|spent on",
        java.util.regex.Pattern.CASE_INSENSITIVE
    );

    @Override
    public void onReceive(Context context, Intent intent) {
        if ("android.provider.Telephony.SMS_RECEIVED".equals(intent.getAction())) {
            Bundle bundle = intent.getExtras();
            if (bundle == null) return;

            Object[] pdus = (Object[]) bundle.get("pdus");
            if (pdus == null) return;

            String format = bundle.getString("format");
            for (Object pdu : pdus) {
                SmsMessage smsMessage;
                if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.M) {
                    smsMessage = SmsMessage.createFromPdu((byte[]) pdu, format);
                } else {
                    smsMessage = SmsMessage.createFromPdu((byte[]) pdu);
                }
                
                String body = smsMessage.getMessageBody();
                String sender = smsMessage.getDisplayOriginatingAddress();

                if (body == null) continue;

                // Simple filter for transaction messages using Regex
                if (TRANSACTION_PATTERN.matcher(body).find()) {
                    Log.d(TAG, "Transaction SMS detected!");

                    // 1. Broadcast for when the app is already in the foreground
                    Intent broadcastIntent = new Intent("com.pixel.reimburse.TRANSACTION_DETECTED");
                    broadcastIntent.putExtra("body", body);
                    broadcastIntent.putExtra("sender", sender);
                    context.sendBroadcast(broadcastIntent);

                    // 2. Start MainActivity to bring it to the foreground (Nudge)
                    if (android.os.Build.VERSION.SDK_INT < android.os.Build.VERSION_CODES.Q || 
                        android.provider.Settings.canDrawOverlays(context)) {
                        
                        Intent activityIntent = new Intent(context, MainActivity.class);
                        activityIntent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_SINGLE_TOP | Intent.FLAG_ACTIVITY_REORDER_TO_FRONT);
                        activityIntent.putExtra("body", body);
                        context.startActivity(activityIntent);
                    } else {
                        NotificationHelper.showTransactionNotification(context, body);
                    }
                }
            }
        }
    }
}

File: android/app/src/main/java/com/pixel/reimburse/SMSReceiver.kt
package com.pixel.reimburse

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.os.Build
import android.provider.Settings
import android.telephony.SmsMessage
import android.util.Log

class SMSReceiver : BroadcastReceiver() {
    private val TAG = "SMSMonitor"
    private val parser = TransactionParserEngine()

    override fun onReceive(context: Context, intent: Intent) {
        if ("android.provider.Telephony.SMS_RECEIVED" == intent.action) {
            val bundle = intent.extras ?: return
            val pdus = bundle.get("pdus") as Array<*>? ?: return
            val format = bundle.getString("format")

            for (pdu in pdus) {
                val smsMessage = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                    SmsMessage.createFromPdu(pdu as ByteArray, format)
                } else {
                    SmsMessage.createFromPdu(pdu as ByteArray)
                }

                val body = smsMessage.messageBody ?: continue
                val sender = smsMessage.displayOriginatingAddress ?: "Unknown"

                Log.d(TAG, "SMS received from $sender: $body")

                val transaction = parser.parse(sender, "SMS", body, System.currentTimeMillis())
                
                if (transaction != null) {
                    Log.i(TAG, "Transaction detected in SMS: ${transaction.amount} at ${transaction.merchant}")
                    
                    // 1. Notify Capacitor Plugin
                    FinancialNotificationPlugin.onTransactionDetected(transaction)
                    
                    // 2. Show Overlay if it's a debit transaction
                    if (transaction.type == "debit") {
                        showOverlay(context, transaction)
                    }
                }
            }
        }
    }

    private fun showOverlay(context: Context, transaction: ParsedTransaction) {
        if (Settings.canDrawOverlays(context)) {
            val intent = Intent(context, OverlayService::class.java).apply {
                putExtra("amount", transaction.amount)
                putExtra("merchant", transaction.merchant)
                putExtra("appName", transaction.appName)
                putExtra("timestamp", transaction.timestamp)
                putExtra("rawText", transaction.rawText)
                addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
            }
            context.startService(intent)
        }
    }
}
